### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by BenjaminStuermer**

### CODE REVIEW

File: src/home_audit.ts [hunk 1]
```diff
@@ -0,0 +1,807 @@
...
+}
+
+/**
+ * Check that the HVAC system is of a valid year
+ */
+const MIN_SYSTEM_YEAR = 1970
+const MAX_SYSTEM_YEAR = (new Date()).getFullYear() // this year
+function checkSystemYear({year}:HeatingSystem|CoolingSystem, name:'heating'|'cooling', index:number): void
+{
+    if(year === undefined || MIN_SYSTEM_YEAR > year || year > MAX_SYSTEM_YEAR) {
```
Comment (thepriestscottage) [inline src/home_audit.ts:678]: This check is incorrect; `year` is nullable in the schema (in favor of user entered efficiency).  If year isn't provided though, this error will always return.

File: src/home_audit.ts [hunk 2]
```diff
@@ -0,0 +1,807 @@
...
+{
+    if(year && (1972 > year || (new Date()).getFullYear() < year)) {
+        addErrorMessage(`systems/domestic_hot_water/year`, `Invalid year, must be between 1972 and ${(new Date()).getFullYear()}`)
+    }
+}
+
+/**
+ * Check that the energy factor is valid for the hot water system
+ */
+function checkHotWaterEnergyFactorValid({type, energy_factor}:HotWater): void
```
Comment (thepriestscottage) [inline src/home_audit.ts:764]: Energy Factor is not a part of the multi-family schema, and it looks like the schema handles this range check anyway, so I think this method can be dropped.

Comment (thepriestscottage) [reply src/home_audit.ts:764]: See regression_files/DHW_indirect.json

File: src/home_audit.ts [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+    }
+}
+
+/**
+ * Check that the insulation level is appropriate for the foundation type
+ */
+function checkFoundationLevel(floors:Floor[]): void
+{
+    floors.forEach(({foundation_type, foundation_insulation_level}, index) => {
+        if(foundation_type && isNullOrUndefined(foundation_insulation_level)) {
```
Comment (thepriestscottage) [inline src/home_audit.ts:390]: `foundation_insulation_level` is not required for all home anymore.  See line 880 of _base.schema.json
Test with regression_files/Foundation_type_above_other_unit.json

File: src/home_audit.ts (hunk 1 above)
Comment (BenjaminStuermer) [reply src/home_audit.ts]: Changed this to `if(year && (MIN_SYSTEM_YEAR > year || year > MAX_SYSTEM_YEAR))`

File: src/home_audit.ts (hunk 2 above)
Comment (BenjaminStuermer) [reply src/home_audit.ts:764]: There is an issue here that I'm not certain how to resolve - there is a mismatch between the domestic_hot_water definitions in HomeEnergyScore.type.ts and in _base.schema.js. In HomeEnergyScore.type.ts, the fields are:

```
category
type
fuel_primary
efficiency_method
year
energy_factor
```

But in _base.schema.js, which we validate against, the fields are:

```
category
type
fuel_primary
efficiency_method
year
efficiency
efficiency_unit
```

So basically, our type definition says we use energy_factor, but the schema says we use efficiency and efficiency_unit.

There is a note in the type file that it is generated using json-schema-to-typescript. Should I regenerate the type definition with that script? I'm not sure how to do that, though I'm sure I can figure out out.

Comment (BenjaminStuermer) [reply src/home_audit.ts:764]: I did remove the check and confirmed by testing that the validation is being handled by the schema

Comment (thepriestscottage) [reply src/home_audit.ts]: @BenjaminStuermer - Did I drop the ball on this?  I remember looking into this and finding the that typescript type had not been correctly built in hes-rest.  The issue required a fix the the schema, which I resolved and was recently approved/merged by NREL.  Do you still need the updated type?

Comment (BenjaminStuermer) [reply src/home_audit.ts]: @thepriestscottage I don't remember the details, but I definitely updated the type definitions. I think we're good.

**Comment by BenjaminStuermer** [comment]
@thepriestscottage Given the scope of the changes here, I'll leave it up to you to merge this, but I think it's ready. We'll need dependent projects to update their code, though, since the main script is in a different location now that it is compiled.

**Pull request merged**